### DISCUSSION TITLE
CLI restructure

### DISCUSSION THREAD

**Discussion by 1st1** [geldata/gel#2450] [Pre-RFC]
This topic has been beaten to death, but keeps resurfacing. So far it's been almost always due to me not able to accept the current implementation, but on our Friday call @colinhacks and @vpetrovykh also shared their concerns.

Previous discussion: edgedb/rfcs#14

We do have RFC 1003 about our CLI design, which suggests we can use essentially two entirely different CLI designs: `[entity] [action]` and `[action]-[entity]`.  We've been using the former only for `edgedb server`, but now we're about to start also using it for `edgedb project`. And something tells me this isn't the last time we do this. Essentially, with the current CLI, I can't remember how most of the commands are called, I can't build a mental model of them, because the design is simply incoherent.

So, in my opinion, our CLI is in this weird state:

* some commands use `[entity] [action]` scheme: `edgedb server` and `edgedb project`.
* migrations inconsistently use `[action]-[entity]` : we have `migration-log` and `create-migration`, and RFC 1003 suggests we should also have `migration-status`. 
* we have a whole family of commands `list-[something]` and `describe` that aren't quite useful in the context of CLI (e.g. how often would `list-casts` be used). If needed, piping `\describe Foo` to `edgedb` shouldn't be a problem for most users, but that's a rare need. These commands, however, significantly clutter the `--help` output.

Paul had an argument that our `\commands` in REPL should be consistent with the CLI, e.g. `edgedb --list-modules` and `\list-modules`. The more I think about the more I'm convinced that it's fine if they're *not* consistent. Working in a REPL and with a CLI are somewhat different modes of operation, and are essentially two different DXes we should optimize. A convenient REPL is important, but so is (if not more) a convenient and easy to memorize commands structure for the CLI.

Here's what we have now:

```
SUBCOMMANDS:
    alter-role
    configure                Modify database configuration
    create-database
    create-migration         Create a migration script
    create-superuser-role
    describe
    drop-role
    dump                     Create a database backup
    help                     Prints this message or the help of the given subcommand(s)
    list-aliases
    list-casts
    list-databases
    list-indexes
    list-modules
    list-object-types
    list-ports
    list-roles
    list-scalar-types
    migrate                  Bring current database to the latest or a specified revision
    migration-log            Show all migration versions
    query
    restore                  Restore a database backup from file
    self-upgrade             Upgrade this edgedb binary
    server
    show-status              Show current migration state
```

I propose we have this instead:

```
SUBCOMMANDS:

  dump                       Create a database backup
  restore                    Restore a database backup from file
  configure                  Configure a DB or a server instance

  migration apply            Migrate the database to the latest revision
  migration create           Create a new migration
  migration log              Show the migrations log
  migration status           Show the current migration state

  project init               Initialize a new EdgeDB project
  project status             Show the status of the current project
  project list               List all projects
  project unlink             Clean-up the project configuration

  server status              Status of an instance
  server init                Initialize a new server instance
  server start               Start an instance
  server stop                Stop an instance
  server restart             Restart an instance
  server info                Show server information
  server install             Install edgedb-server
  server uninstall           Uninstall edgedb-server
  server upgrade             Upgrade installations and instances
  server destroy             Destroy a server instance and remove the data stored
  server list-versions       List available and installed versions of the server
  server logs                Show logs of an instance
  server reset-password      Reset password for a user in the instance
  
  role alter                 Alter an existing DB role
  role create                Create a new DB role
  role create-superuser      Create a new DB superuser role
  role list                  List all available DB roles
  role drop                  Drop a DB role
    
  database create            Create a new DB
  database drop              Drop the DB
  
  describe                   Describe the matching DB object
  list                       List matching database objects, 
                             run `edgedb list help` for details

  self upgrade               Upgrade the "edgedb" CLI tool to the latest version
	
  help                       Prints this message
```

Potentially we can hide some of subcommands and advise users to use, say, `edgedb server help` to see all of them.

Few notes:

1. ~I think that `self-upgrade` should be renamed to `upgrade-cli`, or `cli upgrade` if we think we might be adding more CLI-related commands (maybe `edgedb cli uninstall`?)~ @elprans: '''self makes sense to me, because it refers unambiguously to "this tool you're running now". As opposed to some "cli" (which CLI? this CLI?). We took it from Rust (although there it's self update)'''  So let's call it `self upgrade` and later we might implement `edgedb self uninstall`

2. ~I've dropped all `list-` commands. They're pretty much useless in the CLI, but if we ever want them back, we can always add a new group, like `list [WHAT]`.~ `edgedb list-what` are now replaced with `edgedb list what`

3. We don't need to show all subcommands of `role`, `server`, and `db`. Eventually we can hide some of subcommands and suggest the users to run `edgedb server help` to see all of the options.

4. The current production CLI has an empty `DEBUG OPTIONS:` section. We should remove it if it's empty.

### Backwards Compatibility

We can support all current commands, we just not going to show them in `edgedb --help`. We can show a deprecation warning for them and advise users to use the new version, e.g.:

```
$ edgedb create-migration
The `create-migration` command has been deprecated. Use `edgedb migration create` instead.
```

**Comment by vpetrovykh** [comment]
I'd probably propose to change 
```
migrate                    Migrate the database to the latest revision
```
to 
```
migration sync             Migrate the database to the latest revision
```
Or `apply`, or something along those lines, to bring it under the `migration` umbrella. We could have `edgedb migrate` as a shortcut for `edgedb migration sync` or whatever, though.

Basically I prefer `[entity] [action]` scheme to `[action]-[entity]` simply because I'm more likely to have a good idea which entity I need to work on, than the correct verb for the action I have in mind (or, importantly, whether the action is even possible). So I'd go to `edgedb migration --help` to see how to run a migration or to see if I can rollback maybe, etc.? If I already remember the command by heart, it makes no difference, but in terms of searching for the right command, grouping by entity makes it easier to use the `--help`.

**Reply by 1st1** [reply]
`migration sync` works for me.

**Reply by 1st1** [reply]
What are your overall thoughts on the proposal?

**Reply by vpetrovykh** [reply]
Overall, I like it. Better discoverability and less clutter due to rarely used commands appearing at top-level.

**Comment by elprans** [comment]
One thing I like about the current design is that the `server` (and upcoming `project`) are clear sub-tools with their arguments conventions, whereas bare `edgedb <action>` applies to a specific instance.  The new arrangement erases this distinction.  For example, it becomes much harder to understand which commands require connection arguments and which don't.

> I've dropped all `list-` commands. They're pretty much useless in the CLI,

They're far from useless, we already use them extensively in scripts in CI and Docker entrypoints.

> we can always add a new group, like `list [WHAT]`.

So we have both `role list` and `list role`?

**Reply by 1st1** [reply]
>  For example, it becomes much harder to understand which commands require connection arguments and which don't.

Yeah, I understand the argument. So we're debating what's worse: non-straightforward rule for where to put connection args vs non-straightforward and inconsistent naming of subcommands overall.

IMO remembering that `edgedb project` and `edgedb server` don't have connection options is easy (and makes sense), but this is of course just my subjective opinion.

> They're far from useless, we already use them extensively in scripts in CI and Docker entrypoints.

Do you use `list-aliases` or `list-casts` there? We'll probably continue to add more REPL `\list-` shortcuts in the future, I don't imagine all of them being useful in CLI, yet they grab a lot of attention.

> So we have both role list and list role?

While this isn't super elegant, having aliases like that isn't that horrible. So I'm fine with having both, as long as we don't show the entire `list` subcommand tree in `edgedb --help`.

**Reply by vpetrovykh** [reply]
> So we have both `role list` and `list role`?

Sure. Duplication is not as bad as having to recall whether this particular `list` is an "action-entity" (e.g. `list-databases`) or not (e.g. `project list`). I'd say that the command layout should be favoring those who are not yet intimately familiar with the commands.

**Reply by 1st1** [reply]
> whereas bare edgedb <action> applies to a specific instance. 

Also, part of the problem with the current design is that `<action>` in `edgedb <action>` isn't always a verb, sometimes it's a noun. So `edgedb create-migration` and `edgedb migration-log` just don't play nice together. Renaming `edgedb migration-log` to `edgedb show-migration-log` would make it very hard to type -- bad DX.

**Comment by elprans** [comment]
> `  db configure               Modify the DB configuration`

What about system-level configuration?

**Reply by 1st1** [reply]
> What about system-level configuration?

I wasn't sure what `edgedb configure` does. We can keep it just as `edgedb configure`.

**Comment by colinhacks** [comment]
I think this proposal is more aligned with my expectations and provides a better user experience. 

As a fresh pair of eyes, the RFC struck me as being too concerned with optimizing "consistency", both internal consistency and consistency with REPL commands. API design is an art, not a science, and it's easy to logic yourself into user-hostile designs.

> it becomes much harder to understand which commands require connection arguments and which don't

I think it's pretty self-evident based on the naming and nature of the commands which ones are operating on (or in the context of) a particular database and which aren't. I also don't think the average user thinks critically enough about the grammar of a CLI for this to cause confusion that wouldn't already exist. I don't think the departure from `<action>-<object>` to `<group> <command>` will "tip off" the user to the fact that connection options are optional.

Also, given the introduction of `project init`, there are no longer any commands that really "require connection arguments" (from the user's perspective). I think `project init` is likely to be a far greater contributor to the connection option confusion you're alluding to than the command design.  But for the sake of developer experience both `project init` and the new command structure are worth the tradeoff.

**Reply by 1st1** [reply]
> there are no longer any commands that really "require connection arguments" (from the user's perspective)

And on the other hand, if `project init` isn't used, commands will soon require either `-I instance` or `-H host`, meaning that the user will be instantly prompted to add connection options if they forgot to specify them.

**Comment by 1st1** [comment]
I've updated the proposal:

* ~Moved `list-*` commands under the `db` group (@elprans argues they are quite useful)~
* ~Moved `describe` under the `db` group~
* ~`edgedb migrate` -> `edgedb migration sync`~

~Potentially we can move `restore` and `dump` under the `db` group. If we do that, repl `\` commands would just mirror all subcommands from the `db` group.~

* `edgedb migration apply`
* `edgedb list <what>`
* `edgedb describe`
* `edgedb db` -> `edgedb database`
* `cli upgrade` -> `self upgrade`

**Reply by colinhacks** [reply]
Moving `list-*` and `describe` under `db` seems reasonable, though I'm mostly ambivalent about it. I think they're OK at the top level too. 

I'm not a fan of `edgedb migration sync` for a couple reasons:
1. The term "sync" or "synchronize" is used by several ORMs (TypeORM and Sequelize, to name two) to indicate a dangerous/destructive operation that [isn't intended to be used in production](https://stackoverflow.com/questions/65222981/typeorm-synchronize-in-production). I think it has associations in the JS world as an unsafe/development-only pattern so I'm not sure that's a great choice. 
2. Perhaps `edgedb migration apply` is a better alternative but I personally think `edgedb migrate` is hard to beat here. This is another example of overthinking the API and making it worse in pursuit of consistency. Adding `edgedb migration apply` with `edgedb migrate` as an alias is fine, though in my own day-to-day I would certainly preferentially use `edgedb migrate`.

**Reply by 1st1** [reply]
> The term "sync" or "synchronize" is used by several ORMs (TypeORM and Sequelize, to name two) to indicate a dangerous/destructive operation that

Good point. I've reverted to `edgedb migrate`, but let's wait and see what others say.

> Moving list-* and describe under db seems reasonable, though I'm mostly ambivalent about it. I think they're OK at the top level too.

I think having them under the `db` group allows us to exclude them from `--help` potentially, saving space.

**Reply by elprans** [reply]
I don't think we should move anything under `db` other than operations that relate to creation or removal of databases themselves.

**Reply by 1st1** [reply]
I was thinking about grouping `describe` and `list-*` under some group, so that even if the list grows we can exclude them from being shown in `--help` by default.  Maybe group them under `inspect`?

**Reply by 1st1** [reply]
We can have a top level `list` group, just like we discussed on one of the calls:

```
SUBCOMMANDS:

  dump                       Create a database backup
  restore                    Restore a database backup from file
  configure                  Configure a DB or a server instance

  migrate                    Migrate the database to the latest revision
  migration create           Create a new migration
  migration log              Show the migrations log
  migration status           Show the current migration state

  project init               Initialize a new EdgeDB project
  project status             Show the status of the current project
  project list               List all projects
  project unlink             Clean-up the project configuration

  server status              Status of an instance
  server init                Initialize a new server instance
  server start               Start an instance
  server stop                Stop an instance
  server restart             Restart an instance
  server info                Show server information
  server install             Install edgedb-server
  server uninstall           Uninstall edgedb-server
  server upgrade             Upgrade installations and instances
  server destroy             Destroy a server instance and remove the data stored
  server list-versions       List available and installed versions of the server
  server logs                Show logs of an instance
  server reset-password      Reset password for a user in the instance
  
  role alter                 Alter an existing DB role
  role create                Create a new DB role
  role create-superuser      Create a new DB superuser role
  role list                  List all available DB roles
  role drop                  Drop a DB role
    
  db create                  Create a new DB
  db drop                    Drop the DB

  describe                   Describe the matching DB object
  list                       List matching database objects, 
                             run `edgedb list help` for details

  cli upgrade                Upgrade the CLI to the latest version
	
  help                       Prints this message
```

`db create` and `db drop` can become top-level `createdb` and `dropdb`, if we don't foresee the `db` group ever growing past two commands.

**Reply by elprans** [reply]
The earlier `list` group sounded more reasonable to me.  I don't think we need to lump `describe` together with that.

**Reply by 1st1** [reply]
@colinhacks I've updated the proposal (see the top post) after a quick discussion with Elvis in Slack.

**Reply by colinhacks** [reply]
@1st1 looks like the proposal still lists `edgedb migration apply` instead of `edgedb migrate` - intentional?

**Reply by elprans** [reply]
It's intentional per my suggestion. Its exclusion from the `migration` group seems arbitrary.

**Reply by colinhacks** [reply]
Will `edgedb migrate` still be available as an alias? I really think it's preferable from a DX perspective.

**Comment by tailhook** [comment]
I don't see this list any clearer than what we currently have, in particular:
1. Why `dump`, `restore` are top-level, but `create` / `drop` aren't? They all are relatively rarely used, and related to the database.
2. `self update` is not a command in a group, it's a pair of english words. We use dash in that, like `create-superuser` or `reset-password`.

The same goal can be achieved by just regrouping help:
```
Managing roles and access control:
    alter-role               Change role parameters
    drop-role                Delete a role
    create-superuser-role    Create a new role
    list-roles               Display list of roles in the server instance

Database management:
    create-database          Create a new database
    drop-database
    restore                  Restore a database backup from file
    dump                     Create a database backup

Migrations:
    create-migration         Create a migration script
    apply-migrations         Bring current database to the latest or a specified revision
    show-migrations          Show all migration versions
    check-migrations         Show current migration state

Introspection:
    describe                 Describe a named database object
    list-aliases             Display list of aliases defined in the schema
    list-casts               Display list of casts defined in the schema
    list-databases           Display list of databases in the server instance
    list-indexes             Display list of indexes defined in the schema
    list-modules             Display list of modules defined in the schema
    list-object-types        Display list of object types defined in the schema
    list-scalar-types        Display list of scalar types defined in the schema

Project Management:
    init-project         Initialize Project
    unlink-project       Disassociate project with the database
    
Other:
    query                    Execute EdgeQL query
    self-upgrade             Upgrade this edgedb binary
    server                   Manage local server installations
    configure                Modify database configuration
    help                     Prints this message or the help of the given subcommand(s)
```
(Modified migration and project commands to sync style. And we should think more about "Other" this is just a sketch)

Also we can hide some commands and bring more help topics:
```
[..snip...]
Project Management:
    init-project          Initialize Project
    help project            More help on projects

Help Topics:
    help introspection      Display introspection commands (describe, list-types,...)
    help migrations        Display help on migrations
    help server              Server management commands 
```

1. This way help topics will be more discoverable (in your example, it's not clear that you can run `edgedb migration --help` and whether that will say anything about `migrate` top-level command, and overall guide you through how migrations can be used)
2. We don't need to invent artificial groups like `list`, it's just a verb, not a group. Group should be "introspection" so `describe` fits in, but `inspect describe` and `inspect list-roles` is unnecessarily long. Similarly out of group `migrate` (which will mostly be used by CI not end-users) and `describe`.

**Reply by elprans** [reply]
> `self update` is not a command in a group, it's a pair of english words.

We will have `self uninstall` which would make it a proper group, and `self` is a perfect designation for the group.  FWIW, this is what `rustup` uses too (`rustup self update`, `rustup self uninstall`).

**Reply by colinhacks** [reply]
@tailhook I agree that, intellectually, both designs are equally "clear", but that's not necessarily 1:1 correlated with perceived complexity. I think it's likely you underestimate how hyphen-phobic (hyphenphobic?) the JavaScript ecosystem is, as a developer who's more familiar with a low-level/Unix-ey environment.

**Reply by 1st1** [reply]
> Why dump, restore are top-level, but create / drop aren't? They all are relatively rarely used, and related to the database.

Because potentially `dump` will have an ability to dump all or a number of databases. I'm not too against moving them under the `database` group though.

> I don't see this list any clearer than what we currently have, in particular:

The design your proposed looks OK, until you look at `edgedb server` and `edgedb project` (and maybe we'll also have `edgedb cloud`?) that implement an entirely different design. I know we have this "a tool within a tool" explanation for that, but it doesn't make it look less arbitrary or less unorganized.  We basically have a hodgepodge of different CLI designs now, and I want us to fix that.

**Reply by tailhook** [reply]
> @tailhook I agree that, intellectually, both designs are equally "clear", but that's not necessarily 1:1 correlated with perceived complexity. I think it's likely you underestimate how hyphen-phobic (hyphenphobic?) the JavaScript ecosystem is, as a developer who's more familiar with a low-level/Unix-ey environment.

Well, here are yarn dash commands:
```
    - generate-lock-entry / generateLockEntry
    - upgrade-interactive / upgradeInteractive
```
Here are npm dash commands:
```
 add-user
 dist-tag
 dist-tags
 find-dupes
 help-search
 install-test
 run-script
```
So I don't think hit is a big issue. We will not stop using dash names anyway, even Yury's spec has `reset-password` and `server list-versions` and `create-superuser`...

**Reply by tailhook** [reply]
> Because potentially dump will have an ability to dump all or a number of databases. I'm not too against moving them under the database group though.

I think there was agreement that `dump-all` and `dump` are different commands (but I don't see that in RFC)

> The design your proposed looks OK, until you look at edgedb server and edgedb project (and maybe we'll also have edgedb cloud?) that implement an entirely different design. I know we have this "a tool within a tool" explanation for that, but it doesn't make it look less arbitrary or less unorganized. We basically have a hodgepodge of different CLI designs now, and I want us to fix that.

I don't think `project` is such a thing (as I outlined in my original comment). `server` -- yes is a tool within a tool. And perhaps `cloud` too and they are clear enough. And `configure` is ugly formulated for now, and should be restructred anyway in my opinion.

**Reply by tailhook** [reply]
Also things like:
```
edgedb server reset-password
```
Are quite clear, but:
```
edgedb list scalar-types
```
well, does it easy to remember where to put dash and where to put space? I would say it isn't.

But if you are going to come with solution like: "let's allow spaces everywhere", that opens a huge can of worms like, how to deal with these:
```
edgedb list scalar
edgedb list scalar -d mydb types
```

**Reply by 1st1** [reply]
> well, does it easy to remember where to put dash and where to put space? I would say it isn't.

Unless we have more than 2 levels of nesting it should be easy for users. You can essentially have at most 2 spaces between sub commands. So it's always `edgedb[space]server[space]x-y` and never `edgedb[space]server[space]x[space]y`. Things like these are learned almost subconsciously as long as there's structure.

The current design, however, forces me to accept that `edgedb server` and `edgedb project` are different from `edgedb migration-`, and I'm not convinced that this inconsistency is somehow more acceptable.

**Reply by 1st1** [reply]
> I think there was agreement that dump-all and dump are different commands (but I don't see that in RFC)

@elprans makes an argument to have just one `dump` command that can dump an individual DB, a number of DBs, or all of them.  He has a point. If we have `dump` and `dumpall`, then which one of them would be able to dump just a few selected DBs?  If it's just one `dump` command then it's clear what `dump -d foo -d bar` and `dump --all` do.

**Reply by 1st1** [reply]
> I don't think project is such a thing (as I outlined in my original comment). server -- yes is a tool within a tool. 

Then what you propose, IMO, is strictly worse than both what we have now and what this RFC proposes.

One argument that @elprans keeps repeating is that `edgedb`, currently, has two modes: a tool within a tool mode (`edgedb server` and `edgedb project`) -- these don't need connection options; and all other commands are related to DB management.

What you say is that somehow `project` isn't a tool within a tool, so it must be related to DB management just like `create-role` -- but it isn't. So IMO project has to be a separate tool. And it already has like 4 subcommands defined in its RFC.

**Reply by tailhook** [reply]
> Unless we have more than 2 levels of nesting it should be easy for users. You can essentially have at most 2 spaces between sub commands. So it's always edgedb[space]server[space]x-y and never edgedb[space]server[space]x[space]y. Things like these are learned almost subconsciously as long as there's structure.

This has to be taught explicity. It's not obvious that there can't be `cloud-server` (or whatever) command group. Or `dump-all` top-level command. And `configure` command already ruins this. And I'm not sure we want to put this into a tutorial, so how do you propose to teach this?

**Reply by tailhook** [reply]
> What you say is that somehow project isn't a tool within a tool, so it must be related to DB management just like create-role -- but it isn't. 

I don't think so. We have `edgedb migration-log --from-files` which does not interact with the database at all, this doesn't mean we should move it to `edgedb files migration-log`.

**Reply by 1st1** [reply]
> This has to be taught explicity. 

Yeah, in the proposal we just show all commands spelled out. So I assume a first time user of `edgedb server reset-password` will see it first in `--help`.

I'm not saying that your argument is invalid though, I'm just guessing that it will be a relatively tiny group of people who make this mistake repeatedly. And if they make it only once that is acceptable price to pay.

**Comment by tailhook** [comment]
Another thing to consider is slash commands. Currently they are the same. For example `\list-scalar-types` is equivalent to `edgedb list-scalar-types`.

But if we split them, it might be challenging to make hinting and completion clear enough, given that:
* `\l` is an alias for `\list databases`...
* but `\list` is an alias for a group of commands
* but again `\li` is an alias for `\list indexes`
* and `\lt` is an alias for `\list scalar-types` and `\lT` is an alias for `\list object-types`
* and technically `\list -flags scalar-types` should be working if we obey command-line semantics (or we should write special code to prohibit and hint that, either way it's extra work and bugs).

**Reply by fmoor** [reply]
Why does the cli need to be the same as the repl? I personally don't expect them to be the same. They seem like very different things to me.

**Reply by 1st1** [reply]
Same. I don't think this argument has a lot of weight to it. It's good when everything is consistent, but it's clearly not possible. What we have now is inconsistent in one way, what I propose here is going to be inconsistent for the repl<>cli naming. IMO that's OK in this very case.

**Reply by tailhook** [reply]
> Why does the cli need to be the same as the repl? I personally don't expect them to be the same. They seem like very different things to me.

Because it's hard to remember:
```
$ edgedb list scalar-types
edgedb> \list-scalar-types
```
Arbitrary placed dashes. Also, if we go with the difference we will have to parse both ways anyway and hint that for user:
```
edgedb> \list scalar-types
Unknown command `\list`, try `\list-scalar-types`
$ edgedb list-scalar-types
Unknown command `list-scalar-types` try `edgedb list scalar-types`
```

**Reply by 1st1** [reply]
> Because it's hard to remember:

That implies that someone

a) uses `edgedb list scalar-types`
b) uses full version of `\list-scalar-types` in the REPL (I use shortcuts)

**Reply by tailhook** [reply]
I do both. I still can't remember which one is `\lt` and which one is `\lT`. With completion it isn't hard (it would be even faster if we don't have shortcuts that match prefixes of `\list`, but since `list` is such a frequently used word in programming, I'm typing it non-noticeably anyway).

Also note that `\lT` is hard to suggest to someone in chat or elsewhere in text, because it's hardly distinguishable from `\IT` (depends on font, of course). So I always prefer to write full command in any text communication (it also helps that anyone understands what command does at a glance).

**Reply by tailhook** [reply]
As to why I use `edgedb list-scalar-types`/`list-object-types`, so far it's mostly for finding what's in the database (i.e. whether I'm connecting to a correct one, and whether it is migrated,...). This is running from shell, because I don't need repl for anything else at the moment, and with shell completion of commands it's easy enough. Pipelining doesn't have completion.

I also like running commands instead of using repl because that makes rebuild and test cycle of CLI easier, but that's probably an irrelevant use case. I'm pretty sure users will find other use cases, though.

**Comment by tailhook** [comment]
Another tricky case is `\describe-schema` (https://github.com/edgedb/edgedb-cli/issues/327). It can't go as `describe schema` because `describe` is not a group of commands. But `edgedb describe-schema` will break "no dashes in the first argument" rule which is there now.
*(and yes I can easily come up with use cases for that not in repl)*

**Reply by elprans** [reply]
Perhaps it could just say `\schema` then?

**Reply by 1st1** [reply]
I wouldn't want to waste `edgedb schema` CLI command on this.

**Reply by tailhook** [reply]
Apart from "wasting". It also adds to inconsistency of the introspection group. The most consistent would be `list schema`, but it's a bit weird (although not very bad).

**Reply by 1st1** [reply]
Maybe we just add a top-level group for all introspection-related commands?

```
edgedb inspect list-types
edgedb inspect describe xxx
edgedb inspect describe-schema
edgedb inspect list-modules
```

Alternative names: "intro", "print", "repl", put it under the "db" group.

**Comment by tailhook** [comment]
Additional thoughts after recent chat with @elprans...

If the intention is to provide enough help in tab-completion, we can add `migrations-guide` (or `migration-help` or `migration-tutorial`) commands. So when you complete:
```
$ edgedb migr<tab>
migrate     migration-tutorial
```
In better shells (e.g. zsh) it can even look like:
```
$ edgedb migr<tab>
migrate          -- execute migrations
migration-guide  -- guide on using migrate, create-migration, migration-log... commands
```

**Reply by 1st1** [reply]
I agree that completion is slightly better with `-` as you can see the full list of options. That, however, is just one improvement. The idea of this proposal is to make building a mental model of our commands tree easier. And while there, the idea for our CLI is to look somewhat familiar and at home; `<tool> <group> <action>` is an established approach that I've seen implemented in numerous places and this is why I think it's the winner for us.

**Comment by tailhook** [comment]
Another option, is to reverse the verb-subject to subject-verb:
```
migration-create
role-create
object-types-list
```
(not sure whether it should be plural or not, maybe have both aliases)

The idea is that `migr<tab>` or `role<tab>` still works. But we you don't have to remember whether it `inspect list-object-types` or `info list-object-types`, and whether this is `role create` or `auth create-role`.

**Reply by 1st1** [reply]
I think this just looks weird; the common approach here is to replace the dashes with spaces. Also not everything revolves around code completion, but even if it does, typing `edgedb mig<tab>` -> `edgedb migration<tab>` -> `edgedb migration c<tab>` -> `edgedb migration create` works just fine.

Spaces make namespacing more apparent which leads to a simpler mental model, where commands are neatly compartmentalized based on their functionality. This is something that I feel pretty strongly about, personally, so I don't think I can convince myself to believe that `migration-create` is better than `migration create` :/

**Reply by tailhook** [reply]
When we're speaking about migration only, I can agree with you. but "namespacing more apparent which leads to a simpler mental model" breaks on every second command. I've put numerous examples, like is it `database list` or `introspection list-database`, and so forth.

**Reply by 1st1** [reply]
> When we're speaking about migration only, I can agree with you.

Great. We then have `edgedb migration`, `edgedb server`, `edgedb project`, and will maybe have `edgedb cloud`. At which point we just have to be consistent and employ the `edgedb <group>` approach universally. And I think I'm repeating myself over and over again here, but I just can't see why any potential advantage of theoretically better completion is worth having **2 different** CLI conventions **within one tool**.

> I've put numerous examples, like is it database list or introspection list-database, and so forth.

The answer is both, because why not. One can be an "alias" of the other. This is a very minor detail.

**Reply by tailhook** [reply]
> Great. We then have edgedb migration, edgedb server, edgedb project, and will maybe have edgedb cloud. 

Actually, that's not what I'm saying.

> The answer is both, because why not. One can be an "alias" of the other. This is a very minor detail.

Aliases like that only add to confusion. Because if you read in the tutorial/script command in some other form that you used to you  have to look up whether it's the same thing or not. And now you remember two things :(

(aliases like `migration-log`/`migrations-log` are fine, because they are save you one typo and don't introduce a new thing to remember)

Other than that, yes, we're going in circles. We have already discussed multiple times that it's hard to find out "one non-confusing grouping" of commands, so I don't what to create confusing ones.

**Comment by tailhook** [comment]
Another thing we can do to make `create-migration` more convenient, is to just make `edgedb migrate` an ultimate migration workflow. I.e. it should start interactive create migration workflow right after applying migrations if there are any changes (and and ask to apply it right afterwards). So `create-migration` essentially becomes "plumbing" command.
